docs/source/package-spec.rst

```diff
@@ -355,14 +364,191 @@ packages that this project depends on.
 
-Object Definitions
-------------------
+Definitions
+-----------
 
 Definitions for different objects used within the Package. All objects
 allow custom fields to be included. Custom fields **should** be prefixed
 with ``x-`` to prevent name collisions with future versions of the
 specification.
 
+The *Link Reference* Object
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A link reference object has the following key/value pairs. All link
+references are assumed to be associated with some corresponding bytecode.
+
+Offsets: ``offsets``
+^^^^^^^^^^^^^^^^^^
+
+The ``offsets`` field is an array of integers, corresponding to each of the
+start positions where the link reference appears in the bytecode.
+Locations are 0-indexed from the beginning of the binary representation of
+the corresponding bytecode.  This field is invalid if it references a position
+that is beyond the end of the bytecode.
+
+-  Required: Yes
+-  Type: Array
+
+Length: ``length``
+^^^^^^^^^^^^^^^^^^
+
+The ``length`` field is an integer which defines the length in bytes
+of the link reference. This field is invalid if the end of the defined
+link reference exceeds the end of the bytecode.
+
+-  Required: Yes
+-  Type: Integer
+
+Name: ``name``
+^^^^^^^^^^^^^^
+
+The ``name`` field is a string which **must** be a valid *Identifier*.
+Any link references which **should** be linked with the same
+*link value* **should** be given the same name.
+
+-  Required: No
+-  Type: String
+-  Format: **must** conform to the *Identifier* format.
+
+
+The *Link Value* Object
+~~~~~~~~~~~~~~~~~~~~~~~
+
+A *Link Value* object is defined to have the following key/value pairs.
+
+.. _offset-offset-1:
+
+Offsets ``offsets``
+^^^^^^^^^^^^^^^^^
+
+The ``offsets`` field defines the locations within the corresponding bytecode
+where the ``value`` for this *link value* was written.  These locations are
+0-indexed from the beginning of the binary representation of the
+corresponding bytecode.
+
+-  Required: Yes
+-  Type: Integer
+-  Format: Array of integers, where each integer **must** conform to
+   all of the following:
+
+   -  be greater than or equal to zero
+   -  strictly less than the length of the unprefixed hexidecimal
+      representation of the corresponding bytecode.
+
+Type ``type``
+^^^^^^^^^^^^^
+
+The ``type`` field defines the ``value`` type for determining what is encoded
+when *linking* the corresponding bytecode.
+
+-  Required: Yes
+-  Type: String
+-  Allowed Values:
+
+   -  ``'literal'`` for bytecode literals
+   -  ``'reference'`` for named references to a particular *contract instance*
+
+Value ``value``
+^^^^^^^^^^^^^^^
+
+The ``value`` field defines the value which should be written when
+*linking* the corresponding bytecode.
+
+-  Required: Yes
+-  Type: String
+-  Format: determined based on ``type``:
+
+    Type ``literal``
+      For static value literals (e.g. address), value **must** be a
+      ``'0x'``-prefixed hexadecimal string with representing a whole number of
```

Comment:
I don't think the "representing a whole number of bytes" part should be here, but rather at the location where we define validity for hex encoded bytes.  Then we can just specifically say that the value must be a valid hex encoded byte string.